urlce7759b8da_Palachsuper_EBA4_tgz-pJ8-json_TestJ8/src/terminal/tools/HexTools.java

```diff
@@ -0,0 +1,163 @@
+package terminal.tools;
+
+import java.io.UnsupportedEncodingException;
+
+/**
+ * 16-bit CRC, cyclic redundancy checks, computed from scratch.
+ *
+ * @author Roedy Green, Canadian Mind Products
+ * @version 1.0 2009-01-01 - initial version
+ * @since 2009-01-01
+ */
+public final class HexTools
+{
+    /**
+     * generator polynomial
+     */
+    public static final int poly = 0x1021;/* x16 + x12 + x5 + 1 generator polynomial */
+    /**
+     * scrambler lookup table for fast computation.
+     */
+    public static final int[] crcTable = new int[256];
+
+    static
+    {
+        // initialise scrambler table
+        for ( int i = 0; i < 256; i++ )
+        {
+            int fcs = 0;
+            int d = i << 8;
+            for ( int k = 0; k < 8; k++ )
+            {
+            	fcs = ( ( ( fcs ^ d ) & 0x8000 ) != 0 ) ? ( fcs << 1 ) ^ poly : ( fcs << 1 );
+                d <<= 1;
+                fcs &= 0xffff;
+            }
+            
+            crcTable[ i ] = fcs;
+        }
+    }
+
+    /**
+     * Calc 16-bit CRC with CCITT method.
+     *
+     * @param ba byte array to compute CRC on
+     *
+     * @return 16-bit CRC, unsigned
+     */
+    public static int crc16( byte[] ba ) throws UnsupportedEncodingException
+    {
+        int work = 0x0000;
+
+        for ( byte b : ba )
+        {
+            // xor the next data byte with the high byte of what we have so far to
+            // look up the scrambler.
+            // xor that with the low byte of what we have so far.
+            // Mask back to 16 bits.
+            work = ( crcTable[ ( b ^ ( work >>> 8 ) ) & 0xff ] ^ ( work << 8 ) ) & 0xffff;
+        }
+        
+        return work;
+    }
+
+    public static byte[] fromHexString(String encoded) {
+		if ((encoded.length() % 2) != 0)
+			encoded = "0"+encoded;
+            //throw new IllegalArgumentException("Input string must contain an even number of characters");
+
+        final byte result[] = new byte[encoded.length() / 2];
+        final char enc[] = encoded.toCharArray();
+        
+        for (int i = 0; i < enc.length; i += 2) {
+            StringBuilder curr = new StringBuilder(2);
+            curr.append(enc[i]).append(enc[i + 1]);
+            result[i / 2] = (byte) Integer.parseInt(curr.toString(), 16);
+        }
+
+        return result;
+    }
+	
+	public static String getHexString(byte[] b, int len)
+	{
+		String result = "";
+
+		for (int i=0; i < len; i++)
+			result += Integer.toString( ( b[i] & 0xff ) + 0x100, 16).substring( 1 );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;b&#39; (index: 0 | type: byte[). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;len&#39;. The same parameter type is validated here for example: urlce7759b8da_Palachsuper_EBA4_tgz-pJ8-json_TestJ8/src/terminal/tools/MCrypt.java:82. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).